[Feature] Add archiving departments

vd1992 opened the pull request:
🤖 Resolves #13384

## 👋 Introduction

Adds a feature to be able to archive or unarchive (restore) a department

## 🕵️ Details

Field added, mutations added, query updated.
Then added the advanced tools page with dialogs and the status field to the departments table. 

## 🧪 Testing

1. Navigate to `/en/admin/settings/departments`
2. Head to a department, then the advanced tools tabs
3. Exercise archiving then reversing a department
4. Observe the status column updates in the departments table
5. Observe the archived department can no be selected in places like process creation, but existing references to it unaffected


## 📸 Screenshot

![image](https://github.com/user-attachments/assets/b7568129-cb65-4fc2-b115-05e1d1fbe541)

![image](https://github.com/user-attachments/assets/e32bc725-d055-416a-a4cf-bc270899fcb9)

![image](https://github.com/user-attachments/assets/02b652c8-8f8e-4a91-9a2d-848365e71eb3)

![image](https://github.com/user-attachments/assets/03aed9bb-9aeb-4d32-8b45-525ec1d60214)

![image](https://github.com/user-attachments/assets/82b1661e-c6d9-48ae-ae8b-a9ff566c2bf6)


<!-- Add a screenshot (if possible). -->

## 🖥️ Deployment

Run the roles and permission seeder

vd1992 commented on api/app/GraphQL/Mutations/ArchiveDepartment.php:
This is an operation probably not to be used much and by few people, nor is archived probably a precious data field.
So I didn't fret much about a user somehow trying to re-archive a department. In that case wasn't sure if we'd prefer the timestamp updated or the mutation to do nothing. 

Similar situation for unarchiving

mnigh replied:
This is a good question, but I am not sure on the business logic desired. @gobyrne?

vd1992 commented on api/graphql/schema.graphql:
```diff
@@ -1470,7 +1475,9 @@ type Query {
     @canModel(ability: "viewAny")
   department(id: UUID! @eq): Department @find @canModel(ability: "view")
-  departments: [Department]! @all @canModel(ability: "viewAny")
+  departments(where: DepartmentFilterInput!): [Department]!
```
Made the filter input non-nullable otherwise the scopes don't run at all

vd1992 replied:
I don't follow, searching the schema and I don't see `@spread` on filter `where`'s

petertgiles replied:
I don't love this either.  Having to add an empty query parameter everywhere is pretty ugly.  We want this change to be as transparent as possible.  Let me know if you want me to tinker with it.

vd1992 commented on api/tests/Feature/DepartmentTest.php:
```diff
@@ -319,4 +323,85 @@ public function testCanViewPools()
+    public function testArchivingAndUnArchiving()
+    {
+        $testDepartment = Department::factory()->create(['archived_at' => null]);
+        $user = User::factory()->asAdmin()->create();
```
For some reason I was getting authorization issues using the `$this->adminUser` here so I just seeded another admin user 🤷

petertgiles replied:
Let me know if you want me to tinker with this, too.

vd1992 replied:
Somehow fixed 🤷 
Removed it in https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/18cc192af5603407cdb1a74b9df9f11f63930422

NienkeBr commented:
Hi @vd1992, here's the translation

```
{
  "+12z1r": {
    "defaultMessage": "Le ministère a été restauré.",
    "description": "Alert, department restored successfully"
  },
  "4Nx4Xn": {
    "defaultMessage": "Nom du ministère",
    "description": "Label for the department name input"
  },
  "5uPSDe": {
    "defaultMessage": "Restaurer et activer {departmentName} dans la plateforme.",
    "description": "Dialog subtitle "
  },
  "99pUA+": {
    "defaultMessage": "Ce ministère a été archivé le :",
    "description": "Label for archived date"
  },
  "C6P305": {
    "defaultMessage": "Outils avancés",
    "description": "Nav tab label to access advanced tools "
  },
  "Cx9OvY": {
    "defaultMessage": "Restaurez le ministère",
    "description": "Button to open a dialog to access restoring functionality"
  },
  "E8q/Zm": {
    "defaultMessage": "L’archivage de ce ministère le supprimera des options pouvant être sélectionnées dans l’ensemble de la plateforme. Tout ce qui l’utilise déjà ne changera pas.",
    "description": "Descriptive text for department archiving"
  },
  "GkX/KN": {
    "defaultMessage": "Avertissement! Il s’agit d’actions sensibles qui affecteront l’ensemble de la plateforme. Veuillez donc faire preuve d’une extrême prudence si vous modifiez ces paramètres.",
    "description": "Warning that you are making changes of possibly very high impact"
  },
  "H/0zee": {
    "defaultMessage": "Archivez le ministère",
    "description": "Button to trigger department archiving functionality"
  },
  "ICKgvh": {
    "defaultMessage": "Inscrivez le nom complet du ministère pour confirmer cette action.",
    "description": "Sentence explaining how to archive/restore a department"
  },
  "MiyFrV": {
    "defaultMessage": "Archivé",
    "description": "Status is archived"
  },
  "VC9Tuf": {
    "defaultMessage": "Cette action supprimera le ministère des archives et le restaurera dans la plateforme. Tous ses renseignements précédents pourront à nouveau être consultés.",
    "description": "Descriptive text for department restoring"
  },
  "WNgaJf": {
    "defaultMessage": "Restaurez le ministère",
    "description": "Button to trigger department restoring functionality"
  },
  "Yx8RCD": {
    "defaultMessage": "Le ministère a été archivé",
    "description": "Alert, department archived successfully"
  },
  "mWr4O0": {
    "defaultMessage": "La valeur doit correspondre au nom complet du ministère : {name}",
    "description": "Validation text for department name input"
  },
  "oCvKNa": {
    "defaultMessage": "Archiver et désactiver {departmentName} de la plateforme.",
    "description": "Dialog subtitle "
  },
  "qUkDNj": {
    "defaultMessage": "Restauration du ministère",
    "description": "Heading for restoring a department"
  },
  "qhA97E": {
    "defaultMessage": "Archivez le ministère",
    "description": "Button to open a dialog to access archiving functionality"
  },
  "v8putB": {
    "defaultMessage": "Archivage du ministère",
    "description": "Heading for archiving a department"
  }
}

```

vd1992 commented:
@NienkeBr 

Thank-you!

@mnigh 

Okay, done

vd1992 commented on apps/web/src/pages/Departments/components/RestoreDepartment.tsx:
```diff
@@ -0,0 +1,222 @@
+      <div className="mb-7">
+        <span>
+          {intl.formatMessage({
+            defaultMessage: "This department was archived on: ",
```
https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/d18e699d33b24c11031b5fdca3ff7e030d000766

vd1992 commented on apps/web/src/pages/Departments/components/ArchiveDepartment.tsx:
```diff
@@ -0,0 +1,204 @@
+                        intl.formatMessage(
+                          {
+                            defaultMessage:
+                              "Value must match the department's full name: {name}",
```
https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/d18e699d33b24c11031b5fdca3ff7e030d000766

vd1992 commented on apps/web/src/pages/Departments/components/RestoreDepartment.tsx:
```diff
@@ -0,0 +1,222 @@
+                        intl.formatMessage(
+                          {
+                            defaultMessage:
+                              "Value must match the department's full name: {name}",
```
https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/d18e699d33b24c11031b5fdca3ff7e030d000766

vd1992 commented on apps/web/src/pages/Departments/components/ArchiveDepartment.tsx:
```diff
@@ -0,0 +1,204 @@
+      setOpen(false);
+      toast.success(
+        intl.formatMessage({
+          defaultMessage: "Department archived",
```
https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/d18e699d33b24c11031b5fdca3ff7e030d000766

vd1992 commented on apps/web/src/pages/Departments/utils.ts:
```diff
@@ -32,3 +32,20 @@ export const SIZE_SORT_ORDER = [
+    : intl.formatMessage({
+        defaultMessage: "Published",
+        id: "FBSOkb",
+        description: "Title displayed on the Pool table published at column",
```
Yeah sure, extracted them in https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/61debc0283270ff9920ff3d19a3a96fa0ef15a0e

vd1992 commented on api/app/Policies/DepartmentPolicy.php:
```diff
@@ -82,4 +82,14 @@ public function forceDelete(User $user)
+     *
+     * @return \Illuminate\Auth\Access\Response|bool
+     */
+    public function archiveAndUnarchive(User $user)
```
The schema is more readable this way, I don't think we should condense policies that much. If anything, it'd be better to remove unused policy methods.
This mirrors the pool archive and unarchive structure

petertgiles replied:
I agree with Eric.  Our permission matrix and associated checks tend to be very explicit and verbose.  I'd recommend two new permissions and two policy methods.

vd1992 replied:
Per some talking
Added the permission in https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/18cc192af5603407cdb1a74b9df9f11f63930422

vd1992 commented on api/database/migrations/2025_06_16_210054_departments_add_archived_at.php:
I wondered about this, but the thing is we don't seed through `DepartmentFactory`. The departments are seeded by a big json blob pulled from production

NienkeBr commented on apps/web/src/pages/Departments/components/ArchiveDepartment.tsx:
```diff
@@ -0,0 +1,204 @@
+            <p>
+              {intl.formatMessage({
+                defaultMessage:
+                  "Write the department's full name to confirm this action.",
```
Also GitHub

![image](https://github.com/user-attachments/assets/9d114c47-671c-45de-baeb-5363486ddff8)

Or am i misunderstanding the reason why you can't "write" in a form?

vd1992 commented on api/app/GraphQL/Mutations/ArchiveDepartment.php:
Added validation in https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/a127c84e280620dc3ec3c3f38f39a287eba29fc9

Will be hard to hit though, if you try and update something with an archived department the form treats it as null. Forcing you to pick something else (as intended)

![image](https://github.com/user-attachments/assets/0ef58290-215a-40b0-8aa5-403431166ee6)

![image](https://github.com/user-attachments/assets/cce652d9-88d6-4f02-be20-ca143e0ccb43)

![image](https://github.com/user-attachments/assets/56aeb404-0864-4c13-9e98-d800a28cc12e)

Jerryescandon commented:
I dont think the users require a lot of additional guidance here. 
I think it's ok if the field is empty/error when they click to edit the form. 
we can trust users to go in and re-select the right department (since we expect to only delete duplicate ones, not actually remove a department without a backup)

Jerryescandon commented:
Now, as a nice to have, I would prefer if we could instead add context to the static version, like this

![image](https://github.com/user-attachments/assets/98326bd0-ccf0-4232-af0b-4abab2dd1078)

My pitch was that as part of the archiving process, we would rename it by adding + " (Archived)" at the end before hiding it. This way it would show up everywhere as archived, giving users a little context.

vd1992 commented:
Okay, added the name changing in https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/4c161b54dd7f0c02e48f77d301c95f4c91031e1c
Also updated the test for it

![image](https://github.com/user-attachments/assets/d8034034-2ee7-489a-b63f-e7c73b3ea251)

vd1992 commented:
> Could we please use lower case 'a' for archived?

Heh, I forgot the name was being set in mutation/backend. So it was actually a very easy change to make.
https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/e0842e0cd67dbdf91f37a92be082fae78fc721ac

vd1992 commented on api/config/rolepermission.php:
Okay, done

vd1992 commented on apps/web/src/pages/Departments/components/DepartmentTable.tsx:
```diff
@@ -181,15 +193,20 @@ export const DepartmentTable = ({
+  const [{ data, fetching, error }] = useQuery({
+    query: Departments_Query,
+    variables: {
+      withoutArchived: null, // explicit null value results in the Lighthouse directive not doing a thing
```
Yeah, handy otherwise you might expect null to function like false. Or use the default. 
It was also really handy to have otherwise I would've needed to stick with a scope method

vd1992 commented on apps/web/src/pages/Departments/components/ArchiveDepartment.tsx:
```diff
@@ -0,0 +1,204 @@
+  return (
+    <>
+      <div className="flex justify-center sm:justify-start">
+        <Heading level="h2" color="secondary" className="mt-0 mb-7 font-bold">
```
Reasonable, also looks better https://github.com/GCTC-NTGC/gc-digital-talent/pull/13901/commits/ccef376c44053f53694bb07f4729330937c46df6

petertgiles commented on api/tests/Feature/DepartmentTest.php:
```diff
@@ -319,4 +324,111 @@ public function testCanViewPools()
+        // set to false fetches only archived
+        // no use case for the time being
+        $falseResponse = $this->actingAs($this->baseUser, 'api')
+            ->graphQL($query, ['withoutArchived' => false])
```
Thanks, I think this helps.

petertgiles reviewed (approved):
:heart:
